File: sdk/openfeature-python/examples/basic_usage.py
"""
Basic usage example: Experimently OpenFeature Python Provider.

Run from the repository root with the virtual environment activated:
    source venv/bin/activate
    EXPERIMENTLY_API_KEY=<key> python sdk/openfeature-python/examples/basic_usage.py
"""

from __future__ import annotations

import os
import sys
from pathlib import Path

# Import from the source tree when the packages are not installed.
_HERE = Path(__file__).resolve()
for _path in (_HERE.parents[1] / "src", _HERE.parents[2] / "python"):
    sys.path.insert(0, str(_path))

from openfeature import api  # noqa: E402
from openfeature.evaluation_context import EvaluationContext  # noqa: E402

from experimentation_openfeature import ExperimentationProvider  # noqa: E402


def main() -> None:
    # 1. Create and register the provider. Flags are evaluated by the server and cached
    #    per user + flag for cache_ttl seconds.
    provider = ExperimentationProvider(
        api_key=os.environ.get("EXPERIMENTLY_API_KEY", "your-api-key"),
        base_url=os.environ.get("EXPERIMENTLY_API_URL", "http://localhost:8000"),
        cache_ttl=300,
        timeout=10,
    )
    api.set_provider(provider)

    # 2. Get a client.
    client = api.get_client()

    # 3. Build the evaluation context. targeting_key is the platform user_id and is
    #    required; attributes are kept for your own hooks (they are not sent to the
    #    evaluate endpoint).
    ctx = EvaluationContext(targeting_key="user-12345", attributes={"country": "US", "plan": "pro"})

    # 4. Evaluate flags — same API regardless of provider.
    print("dark-mode enabled:", client.get_boolean_value("dark-mode", False, ctx))
    print("checkout-experiment variant:", client.get_string_value("checkout-experiment", "control", ctx))
    print("cart-max-items:", client.get_integer_value("cart-max-items", 10, ctx))
    print("price-multiplier:", client.get_float_value("price-multiplier", 1.0, ctx))
    print("feature-config:", client.get_object_value("feature-config", {}, ctx))

    # 5. Experiments and tracking are outside OpenFeature: use the SDK client underneath.
    assignment = provider.client.get_variant("checkout_flow", "user-12345", {"plan": "pro"})
    print("checkout_flow variant:", assignment)
    provider.client.track("user-12345", "page_view", properties={"page": "/"})


if __name__ == "__main__":
    main()

File: sdk/elixir/examples/basic_usage.exs
# Basic Usage Example — Experimently Elixir SDK
#
# Flag evaluation and experiment assignment are decided by the server through
# the public API; this script shows every public call. Point it at a running
# backend with EXPERIMENTLY_API_URL / EXPERIMENTLY_API_KEY.
#
# Run with:
#   cd sdk/elixir
#   mix deps.get
#   EXPERIMENTLY_API_KEY=<key> mix run examples/basic_usage.exs

alias Experimently.{Assignment, BatchResult, FlagEvaluation}

# -------------------------------------------------------------------
# 1. Start the client
# -------------------------------------------------------------------
IO.puts("Starting Experimently client...")

{:ok, client} =
  Experimently.start(
    base_url: System.get_env("EXPERIMENTLY_API_URL", "http://localhost:8000"),
    api_key: System.get_env("EXPERIMENTLY_API_KEY", "your-api-key-here"),
    # Seconds a successful evaluation/assignment is reused
    cache_ttl: 300,
    # HTTP timeout in milliseconds
    timeout: 10_000,
    # Keep up to 1000 entries in the cache
    max_cache_size: 1_000
  )

IO.puts("Client started: #{inspect(client)}\n")

# -------------------------------------------------------------------
# 2. Evaluate a feature flag (GET /api/v1/feature-flags/evaluate/{key}?user_id=...)
# -------------------------------------------------------------------
user_id = "user-#{System.unique_integer([:positive])}"

IO.puts("Evaluating feature flag 'dark-mode' for user #{user_id}...")

case Experimently.evaluate_flag(client, "dark-mode", user_id) do
  {:ok, %FlagEvaluation{enabled: true, config: config}} ->
    IO.puts("  => Enabled for this user. Config: #{inspect(config)}")

  {:ok, %FlagEvaluation{enabled: false}} ->
    IO.puts("  => Disabled for this user — showing default light mode")

  {:error, {:auth_error, 401}} ->
    IO.puts("  => Authentication failed — check your API key")

  {:error, {:api_error, 404, _body}} ->
    IO.puts("  => Flag not found or not ACTIVE — using default")

  {:error, {:network_error, reason}} ->
    IO.puts("  => Network error: #{inspect(reason)} — using default")

  {:error, reason} ->
    IO.puts("  => Error: #{inspect(reason)} — using default")
end

# Boolean shorthand: false on any failure.
IO.puts("  feature_enabled?: #{Experimently.feature_enabled?(client, "dark-mode", user_id)}\n")

# -------------------------------------------------------------------
# 3. Get an experiment assignment (POST /api/v1/tracking/assign — sticky on the server)
#    User attributes are sent as `context` for targeting rules.
# -------------------------------------------------------------------
IO.puts("Getting experiment assignment for 'checkout-flow-experiment'...")

case Experimently.get_assignment(client, "checkout-flow-experiment", user_id, %{
       plan: "pro",
       country: "US"
     }) do
  {:ok, %Assignment{is_control: true}} ->
    IO.puts("  => User in CONTROL group — showing existing checkout")

  {:ok, %Assignment{variant_name: variant, configuration: configuration}} ->
    IO.puts("  => User in variant #{variant} — configuration: #{inspect(configuration)}")

  {:error, {:api_error, 404, _body}} ->
    IO.puts("  => Experiment not found or not ACTIVE — showing default checkout")

  {:error, reason} ->
    IO.puts("  => Assignment error: #{inspect(reason)} — showing default checkout")
end

IO.puts("")

# -------------------------------------------------------------------
# 4. Track events (fire-and-forget, always :ok)
# -------------------------------------------------------------------
IO.puts("Tracking events...")

# With a key: one POST /api/v1/tracking/track attributed to the experiment.
:ok =
  Experimently.track(client, "purchase_completed", user_id, %{order_id: "ord-12345"},
    experiment_key: "checkout-flow-experiment",
    value: 99.99
  )

IO.puts("  => Tracked: purchase_completed (experiment_key + value)")

# With a flag key.
:ok =
  Experimently.track(client, "theme_toggled", user_id, %{to: "dark"},
    feature_flag_key: "dark-mode"
  )

IO.puts("  => Tracked: theme_toggled (feature_flag_key)")

# Without a key: fanned out through POST /api/v1/tracking/batch to every
# experiment the user was assigned to and every flag evaluated for the user
# through this client. Nothing cached -> nothing sent.
:ok = Experimently.track(client, "page_viewed", user_id, %{page: "checkout"})
IO.puts("  => Tracked: page_viewed (fanned out to cached assignments + flags)")

# Several events at once (synchronous, max 100 per request).
case Experimently.track_batch(client, [
       %{event_name: "click", user_id: user_id, experiment_key: "checkout-flow-experiment"},
       %{event_name: "click", user_id: user_id, feature_flag_key: "dark-mode"}
     ]) do
  {:ok, %BatchResult{success_count: ok, failure_count: failed}} ->
    IO.puts("  => Batch: #{ok} accepted, #{failed} failed")

  {:error, reason} ->
    IO.puts("  => Batch error: #{inspect(reason)}")
end

IO.puts("")

# -------------------------------------------------------------------
# 5. Cache helpers
# -------------------------------------------------------------------
IO.puts("Cached for #{user_id}:")
IO.puts("  assignments:     #{inspect(Experimently.assignments(client, user_id))}")
IO.puts("  evaluated flags: #{inspect(Experimently.evaluated_flags(client, user_id))}")
IO.puts("")

# -------------------------------------------------------------------
# 6. Direct hash inspection (cross-SDK compatibility check; utility only)
# -------------------------------------------------------------------
IO.puts("Cross-SDK hash check:")
h = Experimently.Evaluator.hash_user("user-123", "my-flag")
expected = 0.6927449859213084
diff = abs(h - expected)
IO.puts("  hash_user(\"user-123\", \"my-flag\") = #{h}")
IO.puts("  expected:                           #{expected}")
IO.puts("  parity: #{if diff < 1.0e-10, do: "OK (matches all SDKs)", else: "MISMATCH!"}")

IO.puts("")

# -------------------------------------------------------------------
# 7. Stop the client (waits for in-flight track requests)
# -------------------------------------------------------------------
IO.puts("Stopping client...")
Experimently.stop(client)
IO.puts("Done.")
